### ISSUE TITLE
Which query boolean operator paradigm should be used?

### ISSUE THREAD

**Issue by Relequestual** [ga4gh-discovery/ga4gh-case-discovery#9]
I'd like to discuss the pros and cons of using the two proposed paradigms for representing boolean operators.

I put forward the [suggestion](https://gist.github.com/Relequestual/65c0446944519a66f8562d02b3cb4c86#query-operators) to use operators as component keys, allowing for nesting, while assuming `AND` if none are specified default.

Anthony Brookes put forward the way used by Cafe Variome where components are given IDs, and the IDs are then combined using boolean logic to formulate the query.

I'll reserve the next comment for filling in my arguments for and against my proposal. Hopefully Anthony (or someone else) can do the same for the altenrative approach.

**Comment by Relequestual** [comment]
Well looks like this wasn't needed as some discussion was had via calls. See further comments.

**Comment by fschiettecatte** [comment]
I prefer the Cafe Variome approach with splitting the components and boolean logic (which is why I put it into the spec I cobbled together), I think it makes the JSON more readable to humans. Also it makes it easier for sites that can't or don't want to implement boolean logic.

**Comment by harindra-a** [comment]
Was this what Tony was referring to today as well? (regarding boolean logic discussion with Ben)

**Comment by Relequestual** [comment]
@harindra-a yes

---
> Also it makes it easier for sites that can't or don't want to implement boolean logic.

@fschiettecatte If that's the case, how would you expect a service to respond if a request uses boolean logic and they do not support it?

**Comment by fschiettecatte** [comment]
With an http status error and a message.

**Comment by Relequestual** [comment]
OK. I think I can live with a separated boolean logic section in the JSON, IF we can use an existing standard for referencing the components, rather than abritrary naming (which requires an extra field for every component!).

In stead, I suggest we use JSON Pointer (RFC6901)
There are plenty of good implementations, and it's supe easy.
For example:
```javascript
var jsonpointer = require('jsonpointer');
var obj = { foo: 1, bar: { baz: 2}, qux: [3, 4, 5]};

jsonpointer.get(obj, '/foo');     // returns 1
jsonpointer.get(obj, '/bar/baz'); // returns 2
jsonpointer.get(obj, '/qux/0');   // returns 3
jsonpointer.get(obj, '/qux/1');   // returns 4
jsonpointer.get(obj, '/qux/2');   // returns 5
jsonpointer.get(obj, '/quo');     // returns undefined
```

I want to avoid people having to write their own way to work out what's being referenced.
Lower the barrier to entry!

**Comment by fschiettecatte** [comment]
Are you thinking of embedding the JSON Pointer to the component in the JSON document itself, that means you need to know the location before you create the JSON document?

**Comment by Relequestual** [comment]
Not necessarily. You can append your JSON components at the same time as compiling the json pointer values.

Alternativly, we could use JSON Path, and then you wouldn't need to know up front.
You could have abitrary naming, but it's more computationally expensive. With JSON Path, you can specify a "query" which can target deep, and look for specific values. This way you could have abritrary naming too, although I'd rather not.

My preference would be to use JSON Pointer, which is simpler. But if it prooves too difficult to implement the use of it, we could use JSON Path.

Does this sound agreeable?

**Comment by fschiettecatte** [comment]
In a word no, I need more time to think about this and I need others to chime in on this.

**Comment by Relequestual** [comment]
In essence, my concern is, I don't want to write something to manually deconstruct a string like `((D AND E AND F) AND K) OR ((D AND E AND F) AND N) OR ((A AND B AND C AND (J AND H)) AND G AND H) OR ((A AND B AND C AND (J AND H)) AND G AND J)`.
If that information can be conveyed in a format which can be decoded by a standard, which has implementations, I find that prefereable.

**Comment by colinveal** [comment]
Hi, just jumping on this following the call yesterday. Could we replace the 'componentID' with a JSON Pointer string referencing the array position of the component

	"query" : {
		"subject" : [
			{"subjectVariant' : [{...},{...}],
			{"gender" : "male"},
			{"disease" : [{...},{...}]}
		],
	"logic" : "( '/query/subject/0' AND ( '/query/subject/1' OR '/query/subject/2' ))"
	}

**Comment by Relequestual** [comment]
(Ignoring the fact that nesting components is still up for debate).

I think this is close, but it doesn't avoid the need to parse the string to determin the logical structure. I want to avoid this at all costs.
As an alternative, this would be my suggestion:

```
{
  "query": {
    "subject": [
      {
        "subjectVariant": [{},{}],
        "gender": "male",
        "disease": [{},{}]
      }
    ],
    "logic": {
      "-AND": [
        "/query/subject/0",
        { 
          "-OR": [
            "/query/subject/1",
            "/query/subject/2"
          ]
        }
      ]
    }
  }
}
```

This way I can traverse down the `logic` object, and when encountering key words, nest or not nest query structure.

It MAY not be as human readable, but I'm more concerned with what makes implementation easier, and to me this clearly does.

**Comment by harindra-a** [comment]
hmm..I think I like @Relequestual 's approach here @colinveal. Parsing stacks of nested parenthesis might become difficult and is tricky. I have bad memories from LISP, sorry..

**Comment by fschiettecatte** [comment]
One benefit of expressing the boolean query in JSON as opposed to a _parenthesied_ string is that you know it will be correct because the JSON has to validate to parse.

**Comment by harindra-a** [comment]
yeah, agree with with @fschiettecatte and @Relequestual. All good point. Parsing parenthesis as strings are going to be a pain to implement for people and error prone. @Relequestual's alternative suggestion works pretty well and seems a compromise @colinveal ? 

Let's lock this down and modify the standard asap please since we are almost on the v0.1.0 freeze..

**Comment by colinveal** [comment]
Sorry been away for a couple of days. We use both methods internally depending on interface and back end, so I'm happy to proceed this way.

**Comment by harindra-a** [comment]
No worries,  thanks @colinveal. And awesome!

@Relequestual,  sounds like we have consensus? Can you or @colinveal  update the standard to reflect this new pattern when one of you gets a chance? 

Nice work guys!

**Comment by Relequestual** [comment]
Proposed solution: https://github.com/ga4gh-discovery/ga4gh-discovery-search/pull/30/commits/5b86630c2879a13f9930b82d4aaf55cc7146274c

Let me know if there are any issues with this, otherwise I'll proceed.

**Comment by Relequestual** [comment]
Merged in as part of #30
